0015-estimator-interface.md

```diff
@@ -0,0 +1,457 @@
+
+### BindingsArray <a name="bindingsarray"></a>
+
+It is common for a user to want to do a sweep over parameter values, that is, to execute the same parametric circuit with many different parameter binding sets. 
+We propose a new class `BindingsArray` that stores multiple sets of parameters, which can be bound to a circuit. 
+For example, if a circuit has 200 parameters, and a user wishes to execute the circuit for 50 different sets of values, then a single instance of `BindingsArray` could represent 50 sets of 200 parameter values. 
+Moreover, we propose that it be array-like, so that, in this example, the `BindingsArray` instance would be one-dimensional and have shape equal `(50,)`. 
+
+The proposed `BindingsArray` object would support, at a minimum, the following constructor examples for a circuit with three input parameters `a`, `b`, and `c`, where we will use `<>` to denote some `array_like` of the specified shape:
+
+```python
+# a single value for each a, b, and c; 0-dimensional, and compatible with current interface
+BindingsArray([0.0, 1.0, 2.0])
+
+# a single value for each a, b, and c; 0-dimensional, and compatible with current interface
+# satisfies user wish to be able to specify values by name
+BindingsArray(kwvals={Parameter("a"): 0.0, Parameter("b"): 1.0, Parameter("c"): 2.0})
+
+# specify all 50 binding parameter sets in one big array
+BindingsArray(<50, 3>) 
+
+# specify bindings separately for each parameter, required if they have different types
+BindingsArray([<50>, <50>, <50>]) 
+
+# include parameter names with the arrays, where parameters can be grouped together in tuples, or supplied separately
+BindingsArray(kwargs={(a, c): <50, 2>, b: <50>}) 
+
+# “args” and “kwargs” can be mixed
+BindingsArray(<50, 2>, {c: <50>}) 
+```
+
+We expect the formal primitive API and primitive implementations to have a strong sense of `BindingsArray`, but we will not demand that users construct them manually because we do not wish to overburden them with types, and we need to remain backwards compatible. This is discussed further in the [Type Coercion Strategy](#type-coercion-strategy) and [Migration Path](#migration-path) sections.
+
+### ObservablesArray <a name="observablesarray"></a>
+
+With the `Estimator`, it is common for a user to want to estimate many observables of a single circuit. 
+For example, all weight-1 and weight-2 Paulis that are adjacent on the connectivity graph. 
+For a one-hundred qubit device, this corresponds to hundreds of unique estimates to be made for a single circuit, noting that for this particular example, on the heavy-hex graph, in the absence of mitigation, only 9 circuits need to be physically run.
+
+We propose a new `ObservablesArray` object, where each element corresponds to a observable the user wants an estimated expectation value of.
+We propose that the internal data model of this object have an element format
+
+```python
+List[Tuple[str, List[int], float]]
```

Comment:
Are we going to expand the alphabet of allowed characters in a `str` representation from (I, X, Y, Z) to also include (0, 1, +, -, l, r) in order to capture state projectors?

Reply:
Yes, I would like to, but my general philosophy with this RFC was to omit features that would be easy to put in a future non-breaking change. If you feel that this does not apply to non-Pauli strings, let me know.

Reply:
Ok, your approach makes sense.